File: HSRadarView.m
//
//  HSRadarView.m
//  CloneData
//
//  Created by 何少博 on 2017/7/1.
//  Copyright © 2017年 shaobo.He. All rights reserved.
//

#import "HSRadarView.h"


@interface HSWaveView : UIView

@property(nonatomic,strong)UIColor *borderColor;

@end

@implementation HSRadarView

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self setDefaultValue];
        [self setup];
    }
    return self;
}

-(instancetype)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setDefaultValue];
        [self setup];
    }
    return self;
}


-(void)setWaveCount:(NSInteger)waveCount{
    _waveCount = waveCount;
    [self setup];
}
-(void)setWaveColor:(UIColor *)waveColor{
    _waveColor = waveColor;
    [self setup];
}
-(void)setWaveWidth:(CGFloat)waveWidth{
    _waveWidth = waveWidth;
    [self setup];
}
-(void)setWaveMaxAlpha:(CGFloat)waveMaxAlpha{
    _waveMaxAlpha = waveMaxAlpha;
    [self setup];
}
-(void)setAnimationDuration:(CGFloat)animationDuration{
    _animationDuration = animationDuration;
    [self setup];
}
-(void)setAnimationOptions:(UIViewAnimationOptions)animationOptions{
    _animationOptions = animationOptions;
    [self setup];
}

-(void)setDefaultValue{
    _waveCount = 5;
    _waveWidth = 40;
    _waveColor = [UIColor blueColor];
    _waveMaxAlpha = 1.0;
    _animationDuration = 6.6;
    _animationOptions = UIViewAnimationOptionCurveEaseOut;
}

-(void)setup{
    for (UIView * view in self.subviews) {
        [view removeFromSuperview];
    }
    
    for (int i = 0; i < _waveCount; i++) {
        HSWaveView * wave = [[HSWaveView alloc]initWithFrame:CGRectMake(0, 0, self.waveWidth, self.waveWidth)];
        wave.center = self.center;
        wave.borderColor = self.waveColor;
        wave.alpha = self.waveMaxAlpha;
        [self addSubview:wave];
    }
}

-(void)startAnimation{
    if (self.subviews.count == 0) {
        [self setup];
    }
    
    NSArray * subViews = self.subviews;
    CGFloat scale = self.bounds.size.width/self.waveWidth;
    
    for (int index = 0; index < subViews.count; index ++) {
        UIView * view = subViews[index];
        CGFloat delay = self.animationDuration/self.waveCount*index;
        [UIView animateWithDuration:self.animationDuration delay:delay options:self.animationOptions|UIViewAnimationOptionRepeat animations:^{
            view.transform = CGAffineTransformScale(view.transform, scale, scale);
            view.alpha = 0.0;
        } completion:nil];
    }
    
}
-(void)stopAnimation{
    for (UIView * view in self.subviews) {
        [view removeFromSuperview];
    }
}

@end




@implementation HSWaveView

-(void)setBorderColor:(UIColor *)borderColor{
    _borderColor = borderColor;
    [self setNeedsDisplay];
}

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
    }
    return self;
}
-(instancetype)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
    }
    return self;
}

-(void)drawRect:(CGRect)rect{
    CGFloat redius = rect.size.width/2;
    UIBezierPath * path = [UIBezierPath bezierPathWithRoundedRect:rect cornerRadius:redius];
    path.lineWidth = 1.0;
    CAShapeLayer * layer = [CAShapeLayer layer];
    layer.path = path.CGPath;
    layer.strokeColor = self.borderColor.CGColor;
    layer.fillColor  = self.borderColor.CGColor;
    [self.layer addSublayer:layer];
}




@end

File: HSRadarView.swift
//
//  HSRadarView.swift
//  CloneData
//
//  Created by 何少博 on 2017/6/21.
//  Copyright © 2017年 Yanqh. All rights reserved.
//

import UIKit




class HSRadarView: UIView {
    fileprivate  var waveWidth:CGFloat = 40
    var waveCount: Int = 5 { didSet { setup() } }
    var waveColor: UIColor = UIColor.blue { didSet { setup() } }
    var waveMaxAlpha: CGFloat = 1.0 { didSet { setup() } }
    var animationDuration: Double = 6.6 { didSet { setup() } }
    var animationOptions: UIViewAnimationOptions = [.curveEaseOut] { didSet { setup() } }
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    func setup() {
        for view in self.subviews {
            view.removeFromSuperview()
        }
        
        for _ in 0..<waveCount {
            let wave = HSWaveView(frame: CGRect(x: 0, y: 0, width: waveWidth, height: waveWidth))
            wave.center = self.center
            wave.borderColor = waveColor
            wave.alpha = waveMaxAlpha
            
            self.addSubview(wave)
        }
    }
    
    func startAnimation() {
        if self.subviews.count == 0 {
            setup()
        }
        let views = self.subviews
        let scale = self.bounds.width / waveWidth
        for (index, view) in views.enumerated() {
            
            let delay = animationDuration / Double(waveCount) * Double(index)
            UIView.animate(withDuration: animationDuration, delay: delay, options: [.repeat, animationOptions], animations: {
                view.transform = view.transform.scaledBy(x: scale, y: scale);
                view.alpha = 0.0
            }, completion: nil)
        }
    }
    func stopAnimation() -> () {
        for view in self.subviews {
            view.removeFromSuperview()
        }
    }
    
}

fileprivate class HSWaveView: UIView {
    
    var borderColor = UIColor.blue { didSet{ setNeedsDisplay() } }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = UIColor.clear
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        backgroundColor = UIColor.clear
    }
    
    override func draw(_ rect: CGRect) {
        
        let radius = rect.width/2
        let path = UIBezierPath(roundedRect: rect, cornerRadius: radius)
        path.lineWidth = 1.0
        
        let layer = CAShapeLayer()
        layer.path = path.cgPath;
        layer.strokeColor = borderColor.cgColor
        layer.fillColor = borderColor.cgColor//UIColor.clear.cgColor
        self.layer.addSublayer(layer)
        
    }
    
    
}



